File: python/src/features/users/get_user.py
"""GetUser slice — fetch a user by id when authenticated."""

from __future__ import annotations

from dataclasses import dataclass

from fastapi import APIRouter, Depends
from pydantic import BaseModel, ConfigDict
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from authentication.dependencies import get_current_subject
from common.cache import AppCache
from common.problem import problem_response
from common.result import Result
from database.models import UserModel
from database.session import get_db
from features.users.cache_keys import UserCacheKeys
from features.users.dependencies import get_app_cache
from features.users.errors import UserErrors

router = APIRouter(prefix="/api/v1/users", tags=["Users"])

_CACHE_TTL_SECONDS = 60.0


@dataclass(frozen=True, slots=True)
class Query:
    id: int


@dataclass(frozen=True, slots=True)
class Response:
    id: int
    name: str
    email: str


class _UserNotFound(Exception):
    pass


class GetUserResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: int
    name: str
    email: str


class Handler:
    def __init__(self, session: AsyncSession, cache: AppCache) -> None:
        self._session = session
        self._cache = cache

    async def handle(self, query: Query) -> Result[Response]:
        cache_key = UserCacheKeys.by_id(query.id)

        async def factory() -> Response:
            model = (
                await self._session.execute(
                    select(UserModel).where(UserModel.id == query.id)
                )
            ).scalar_one_or_none()
            if model is None:
                raise _UserNotFound()
            return Response(id=model.id, name=model.name, email=model.email)

        try:
            response = await self._cache.get_or_set(
                cache_key,
                factory,
                ttl_seconds=_CACHE_TTL_SECONDS,
            )
        except _UserNotFound:
            return Result.failure(UserErrors.not_found())

        return Result.success(response)


def _get_handler(
    session: AsyncSession = Depends(get_db),
    cache: AppCache = Depends(get_app_cache),
) -> Handler:
    return Handler(session, cache)


@router.get("/{user_id}", response_model=None)
async def get_user(
    user_id: int,
    _: int = Depends(get_current_subject),
    handler: Handler = Depends(_get_handler),
) -> GetUserResponse | object:
    result = await handler.handle(Query(id=user_id))
    return result.match(
        on_success=lambda value: GetUserResponse(
            id=value.id,
            name=value.name,
            email=value.email,
        ),
        on_failure=problem_response,
    )

File: dotnet/src/Manager.Api/Features/Users/GetUser.cs
using Manager.Api.Common;
using Manager.Api.Common.Messaging;
using Manager.Api.Database;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Caching.Hybrid;

namespace Manager.Api.Features.Users;

public static class GetUser
{
    public sealed record Query(long Id) : IQuery<Response>;

    public sealed record Response(long Id, string Name, string Email);

    internal sealed class Handler(
        ApplicationDbContext db,
        HybridCache cache) : IQueryHandler<Query, Response>
    {
        public async Task<Result<Response>> Handle(Query query, CancellationToken cancellationToken)
        {
            var response = await cache.GetOrCreateAsync(
                UserCacheKeys.ById(query.Id),
                async ct =>
                {
                    return await db.Users
                        .AsNoTracking()
                        .Where(u => u.Id == query.Id)
                        .Select(u => new Response(u.Id, u.Name, u.Email))
                        .SingleOrDefaultAsync(ct);
                },
                cancellationToken: cancellationToken);

            if (response is null)
            {
                await cache.RemoveAsync(UserCacheKeys.ById(query.Id), cancellationToken);
                return Result.Failure<Response>(UserErrors.NotFound());
            }

            return Result.Success(response);
        }
    }

    public sealed class Endpoint : IEndpoint
    {
        public void MapEndpoint(IEndpointRouteBuilder app) =>
            app.MapGet(
                    "/api/v1/users/{id:long}",
                    async (
                        long id,
                        [FromServices] IQueryHandler<Query, Response> handler,
                        CancellationToken cancellationToken) =>
                    {
                        var result = await handler.Handle(new Query(id), cancellationToken);
                        return result.Match(Results.Ok, CustomResults.Problem);
                    })
                .RequireAuthorization()
                .WithTags(Tags.Users);
    }
}
